### PR TITLE
Add a configurable memory limit to the torrent repository and implement dashmap [#567, #565]

### LINKED ISSUE #565: Use dashmap for torrent in memory storage instead of the std hashmap
DashMap should be significantly faster in concurrent reads and writes compared to the std HashMap.

https://github.com/xacrimon/dashmap
https://github.com/xacrimon/conc-map-bench

### LINKED ISSUE #567: Add a configurable memory limit to the torrent repository
Currently the torrent repository is not bound by an upper limit of torrents or memory usage. When the memory size of the torrent repository exceeds the amount of memory available on the host system, the tracker will [restart](https://github.com/torrust/torrust-tracker/issues/566).

To solve this problem, I want to add a configurable memory limit for the torrent repository. When the torrent repository is full, it should let new torrents replace old low priority torrents. For the first implementation I will base the priority on the time it was last accessed in an announce request.

### Relevant discussions/issues
https://github.com/torrust/torrust-tracker/discussions/499
https://github.com/torrust/torrust-tracker/issues/566

### Needs issues
https://github.com/torrust/torrust-tracker/issues/565

cc @josecelano @da2ce7

### PR DESCRIPTION
**Pull request by mickvandijke**
Closes https://github.com/torrust/torrust-tracker/issues/567
Closes https://github.com/torrust/torrust-tracker/issues/565

### CODE REVIEW

**Comment by mickvandijke** [comment]
Needs rebase and commit cleanup first before reviewing.

File: src/core/torrent/repositories.rs [hunk 1]
```diff
@@ -476,32 +393,47 @@ impl Repository for RepositoryDashmap {
     fn upsert_torrent_with_peer_and_get_stats(&self, info_hash: &InfoHash, peer: &peer::Peer) -> (SwarmStats, bool) {
         let hash = self.torrents.hash_usize(&info_hash);
         let shard_idx = self.torrents.determine_shard(hash);
+        let mut shard = unsafe { self.torrents._yield_write_shard(shard_idx) };
```
Comment (da2ce7) [inline src/core/torrent/repositories.rs:394]: it is good to explain why unsafe code is needed, and why this use is safe

**Comment by josecelano** [comment]
Hi @mickvandijke what is the status of this PR? Maybe:

- Rebase
- And https://github.com/torrust/torrust-tracker/issues/646

By the way, why did you use [deepsize](https://github.com/Aeledfyr/deepsize/)?

I'm just curious, I've found this one https://github.com/str4d/memuse

**Comment by josecelano** [comment]
Hi @mickvandijke I'm closing this PR:

- [The Dashmap was already implemented](https://github.com/torrust/torrust-tracker/blob/develop/packages/torrent-repository/src/repository/dash_map_mutex_std.rs).
- The [memory limit should be implemented for all repos](https://github.com/torrust/torrust-tracker/discussions/789) or [not implemented at all](https://github.com/torrust/torrust-tracker/discussions/825).

**Pull request closed without merging**